File: Capture the Flag/src/main/java/me/yoursole/ctf/Events/BlockBreak.java
package me.yoursole.ctf.Events;

import com.google.common.collect.Maps;
import me.yoursole.ctf.DataFiles.GameData;
import me.yoursole.ctf.DataFiles.Utils;
import org.apache.commons.lang.math.RandomUtils;
import org.bukkit.GameMode;
import org.bukkit.Material;
import org.bukkit.entity.Item;
import org.bukkit.entity.Player;
import org.bukkit.event.EventHandler;
import org.bukkit.event.EventPriority;
import org.bukkit.event.Listener;
import org.bukkit.event.block.BlockBreakEvent;
import org.bukkit.inventory.ItemStack;

import java.util.Collection;
import java.util.Map;

public class BlockBreak implements Listener {

    private static final Map<Material, Material> smeltable = Maps.newHashMap();
    private static final Map<Material, Float> experience = Maps.newHashMap();

    @EventHandler(priority = EventPriority.LOWEST)
    public void onBlockBreak(BlockBreakEvent event) {
        if (!event.isDropItems() || event.getPlayer().getGameMode() == GameMode.CREATIVE) return;
        Collection<ItemStack> drops = event.getBlock().getDrops(event.getPlayer().getInventory().getItemInMainHand());
        if (drops.isEmpty()) return;
        for (ItemStack drop : drops) {
            Material smelted = smeltable.get(drop.getType());
            if (smelted != null) {
                ItemStack stack = new ItemStack(smelted);
/*                if (smelted == Material.COPPER_INGOT) {
                    stack.setAmount(getRandom(2, 3));
                }*/
                stack.setAmount(drop.getAmount());

                event.getPlayer().setTotalExperience((int) (event.getPlayer().getTotalExperience() + experience.getOrDefault(smelted, 0f) * stack.getAmount()));
                drop = stack;
            }
            Utils.doTelekinesis(event.getPlayer(), drop);
        }
        event.setDropItems(false);
    }

    static {
        smeltable.put(Material.RAW_IRON, Material.IRON_INGOT);
        smeltable.put(Material.RAW_GOLD, Material.GOLD_INGOT);
        smeltable.put(Material.RAW_COPPER, Material.COPPER_INGOT);
/*        smeltable.put(Material.IRON_ORE, Material.IRON_INGOT);
        smeltable.put(Material.DEEPSLATE_IRON_ORE, Material.IRON_INGOT);
        smeltable.put(Material.COPPER_ORE, Material.COPPER_INGOT);
        smeltable.put(Material.DEEPSLATE_COPPER_ORE, Material.COPPER_INGOT);
        smeltable.put(Material.GOLD_ORE, Material.GOLD_INGOT);
        smeltable.put(Material.DEEPSLATE_GOLD_ORE, Material.GOLD_INGOT);*/

        experience.put(Material.IRON_INGOT, 0.7f);
        experience.put(Material.COPPER_INGOT, 0.7f);
        experience.put(Material.GOLD_INGOT, 1f);
    }
}

File: src/main/kotlin/me/yoursole/ctf/events/BlockBreak.kt
package me.yoursole.ctf.events

import me.yoursole.ctf.datafiles.GameData
import me.yoursole.ctf.datafiles.Utils.doTelekinesis
import org.bukkit.GameMode
import org.bukkit.Material
import org.bukkit.event.EventHandler
import org.bukkit.event.EventPriority
import org.bukkit.event.Listener
import org.bukkit.event.block.BlockBreakEvent
import org.bukkit.inventory.ItemStack
import kotlin.math.roundToInt
import kotlin.random.Random
import kotlin.random.nextInt

object BlockBreak : Listener {
    @EventHandler(priority = EventPriority.LOWEST)
    fun onBlockBreak(event: BlockBreakEvent) {
        if (GameData.dropLoc != null && event.block.location.distanceSquared(GameData.dropLoc!!) <= 9) return
        if (event.isCancelled || !event.isDropItems || event.player.gameMode == GameMode.CREATIVE) return
        val drops = event.block.getDrops(event.player.inventory.itemInMainHand)
        if (drops.isEmpty()) return
        for (d in drops) {
            var drop = d
            val smelted = smeltable[drop.type]
            if (smelted != null) {
                val stack = ItemStack(smelted)
                if (smelted == Material.COPPER_INGOT) {
                    stack.amount = Random.nextInt(2..3)
                }
                stack.amount = drop.amount
                event.player.giveExp((experience.getOrDefault(smelted, 0f) * stack.amount).roundToInt(), true)
                drop = stack
            }
            event.player.doTelekinesis(drop)
        }
        event.player.giveExp(event.expToDrop, true)
        event.isDropItems = false
        event.expToDrop = 0
    }

    private val smeltable = hashMapOf(
        Material.RAW_IRON to Material.IRON_INGOT,
        Material.RAW_GOLD to Material.GOLD_INGOT,
        Material.RAW_COPPER to Material.COPPER_INGOT

    )
    private val experience = hashMapOf(
        Material.IRON_INGOT to 0.7f,
        Material.GOLD_INGOT to 1f,
        Material.COPPER_INGOT to 0.7f
    )
}